lib/std/io/Reader/test.zig

```diff
@@ -412,3 +412,89 @@ test "readStruct reads packed structs without padding and in correct field order
         },
     }
 }
+
+test "readStruct writeStruct round-trip with packed structs" {
+    var buf: [8]u8 = undefined;
+    var fis = std.io.fixedBufferStream(&buf);
+    const reader = fis.reader();
+    const writer = fis.writer();
+
+    const PackedStruct = packed struct(u64) {
+        a: u16 = 123,
+        b: u16 = 245,
+        c: u16 = 456,
+        d: i13 = -345,
+        e: i3 = 2,
+    };
+
+    const expected_packed_struct = PackedStruct{};
+    try writer.writeStruct(expected_packed_struct);
+    fis.reset();
+    try testing.expectEqualDeep(expected_packed_struct, try reader.readStruct(PackedStruct));
+}
+
+test "readStructEndian writeStructEndian round-trip with packed structs" {
+    var buf: [8]u8 = undefined;
+    var fis = std.io.fixedBufferStream(&buf);
+    const reader = fis.reader();
+    const writer = fis.writer();
+
+    const PackedStruct = packed struct(u64) {
+        a: u13 = 123,
+        b: i7 = -24,
+        c: u20 = 83,
+        d: enum(i24) { val = 3452 } = .val,
+    };
+
+    const expected_packed_struct = PackedStruct{};
+    // round-trip little endian
+    try writer.writeStructEndian(expected_packed_struct, .big);
+    fis.reset();
+    try testing.expectEqualDeep(expected_packed_struct, try reader.readStructEndian(PackedStruct, .big));
+    // round-trip big endian
+    fis.reset();
+    try writer.writeStructEndian(expected_packed_struct, .little);
+    fis.reset();
+    try testing.expectEqualDeep(expected_packed_struct, try reader.readStructEndian(PackedStruct, .little));
+}
+
+test "readStruct a packed struct with endianness-affected types" {
+    const buf = [4]u8{ 0x12, 0x34, 0x56, 0x78 };
+    var fis = std.io.fixedBufferStream(&buf);
+    const reader = fis.reader();
+
+    const PackedStruct = packed struct(u32) { a: u16, b: u16 };
+
+    switch (native_endian) {
+        .little => {
+            try testing.expectEqualDeep(
+                PackedStruct{ .a = 0x3412, .b = 0x7856 },
+                reader.readStruct(PackedStruct),
+            );
+        },
+        .big => {
+            try testing.expectEqualDeep(
+                PackedStruct{ .a = 0x5678, .b = 0x1234 },
```

Comment:
Thanks for the updated tests, this makes the behavior much clearer!

But, it still seems backwards to me.  I'd expect `a` to be `0x1234` here (given `a` is `0x3412` on little endian).  That said, I don't have any experience with packed-bit structures on big endian systems, so I may just have the wrong expectation here ...  In my mental model the packed-into-a-u32 struct `{a: u16, b: u16}` is similar in layout to a two-element `u16` array: the first two bytes are the `a` element (in either endianness) and the 3rd and 4th byte are the `b` element (in either endianness).

Reply:
`packed struct` bit-packs elements into a single backing integer, from the first field in the lowest-value bits to the last field in the highest-value bits.
On little endian the lowest-value component is in the lowest address, while on big endian the lowest-value component is in the highest address. Going by address, fields are already swapped between endiannesses in status-quo.

The reason this test swaps in particular has the fields swapped is that we start off with a `[4]u8` as reference data.
The two lower-value bytes of the underlying `u32` always hold `a`. Depending on which endianness you view this under, these will be at address offsets `0` and `1` for little endian, and at offsets `3` and `2` for big endian.

Reply:
Thanks for the explanation @rohlem, that clarifies things for me.